File: Core/src/cells/stalk.rs
use std::sync::{Arc, Mutex};
use crate::cells::{Cell, Action, Position};
use crate::data::Direction;
use rand::{Rng, SeedableRng};
use crate::cells::air::Air;
use crate::cells::cap::Cap;
use crate::cells::mycelium::Mycelium;
use crate::grid::grid;

pub struct Stalk {
    water: Arc<Mutex<f32>>,
    energy: Arc<Mutex<f32>>,
    dies: Arc<Mutex<u32>>,
    target_y: i32,
    rng: Arc<Mutex<rand::rngs::StdRng>>,
}

impl Stalk {
    pub fn new(target_y: i32) -> Self {
        Self {
            water: Arc::new(Mutex::new(0.6)),
            energy: Arc::new(Mutex::new(0.3)),
            dies: Arc::new(Mutex::new(0)),
            target_y,
            rng: Arc::new(Mutex::new(rand::rngs::StdRng::from_os_rng())),
        }
    }
}

impl Cell for Stalk {
    fn do_action(&self, position: Position) -> Option<Action> {
        let mut water = self.water.lock().unwrap();
        let mut energy = self.energy.lock().unwrap();
        let mut dies = self.dies.lock().unwrap();
        let mut rng = self.rng.lock().unwrap();

        if *energy > 0.0001 { *energy -= 0.0001; } else { *dies += 1; }
        if *water > 0.001 { *water -= 0.001; } else { *dies += 1; }

        if *dies > 0 && *energy > 0.2 && *water > 0.2 {
            *energy -= 0.1;
            *water -= 0.1;
            *dies -= 1;
        }

        if *dies > 50 {
            return Some(Box::new(move || {
                let down_pos = position.get_neighbor(Direction::Down);
                if let Some(cell) = grid::get(down_pos).unwrap().as_any().downcast_ref::<Mycelium>() {
                    *cell.water.lock().unwrap() += 0.1;
                    *cell.energy.lock().unwrap() += 0.1;
                }
                grid::set(position, Arc::new(Air));
            }));
        }

        if position.y > self.target_y && *water >= 0.6 && *energy >= 0.3 {
            let up_pos = position.get_neighbor(Direction::Up);
            if grid::get(up_pos).unwrap().as_any().is::<Air>() {
                *water -= 0.6;
                *energy -= 0.3;

                let target_y = self.target_y;
                return Some(Box::new(move || {
                    grid::set(
                        up_pos,
                        Arc::new(Stalk::new(target_y))
                    );
                }));
            }
        }

        if position.y == self.target_y {
            if *energy > 0.5 && *water > 0.5 {
                let up_pos = position.get_neighbor(Direction::Up);
                if grid::get(up_pos).unwrap().as_any().is::<Air>() {
                    let cap_size = rng.random_range(3..6);
                    let max_dist = Position::new(1, 1).distance_squared(Position::new(cap_size, cap_size));

                    *water -= 0.4;
                    *energy -= 0.4;

                    return Some(Box::new(move || {
                        grid::set(
                            up_pos,
                            Arc::new(Cap::new(position, max_dist))
                        );
                    }));
                }
            }
        }

        None
    }

    fn get_color(&self, _: Position) -> String {
        "#f7f8f9".to_string()
    }

    fn clone_cell(&self) -> Arc<dyn Cell> {
        Arc::new(Self {
            water: Arc::clone(&self.water),
            energy: Arc::clone(&self.energy),
            dies: Arc::clone(&self.dies),
            target_y: self.target_y,
            rng: Arc::clone(&self.rng),
        })
    }
}

File: Mushroom/Cells/Stalk.cs
using System;
using Mushroom.Data;
using Godot;

namespace Mushroom.Ceils;

public class Stalk : ICell
{
    public float Water { get; set; } = 0.6f;
    public float Energy { get; set; } = 0.3f;
    private int dies { get; set; } = 0;

    private Random _rand = new();
    public int TargetY { get; set; } = Grid.GroundLevel - 4;
    
    public Action? Do(Vector2I vector2)
    {
        var down = Grid.GetNeighbor(vector2, Direction.Down);
        if (down is not Stalk && down is not Mycelium)
        {
            dies += 1; 
            Energy = 0;
            Water = 0;
        }

        if (Energy >= 0.0001f)
            Energy -= 0.0001f;
        else
            dies++;
        if (Water >= 0.0001f)
            Water -= 0.0001f;
        else
            dies++;

        if (dies > 0 && Energy > 0.02f && Water > 0.02f)
        {
            Energy -= 0.01f;
            Water -= 0.01f;
            dies--;
        }

        if (dies > 20)
        {
            if (down is Stalk stalkDown)
            {
                stalkDown.Energy += Math.Max(0, Energy);
                stalkDown.Water += Math.Max(0, Water);
                Grid.Set(vector2, new RottingMatter() { BecomeAir = true } );
            }
            else if (down is Mycelium mycelium)
            {
                mycelium.Energy += Math.Max(0, Energy);
                mycelium.Water += Math.Max(0, Water);
                Grid.Set(vector2, new RottingMatter() { BecomeAir = true } );
            }
            else 
                Grid.Set(vector2, new RottingMatter() { BecomeAir = true } );
        }
        
        var up = Grid.GetNeighbor(vector2, Direction.Up);
        if (vector2.Y > TargetY && Water >= 0.6f && Energy >= 0.3f)
        {
            if (up is Air)
            {
                return new Action(() =>
                {
                    Water -= 0.6f;
                    Energy -= 0.3f;
                    
                    Grid.Set(vector2.X, vector2.Y - 1, new Stalk() { TargetY = TargetY });
                });   
            }
        }
        else if (vector2.Y == TargetY)
        {
            if (up is Cap cap && (cap.Water < 0.5f || cap.Energy < 0.5f))
            {
                return new Action(() =>
                {
                    float waterToGive = Math.Min(0.5f - cap.Water, Math.Max(0, Water));
                    cap.Water += waterToGive;
                    Water -= waterToGive;

                    float energyToGive = Math.Min(0.5f - cap.Energy, Math.Max(0, Energy));
                    cap.Energy += energyToGive;
                    Energy -= energyToGive;
                });
            }
            if (up is Air && Energy > 0.5f && Water > 0.5f)
            {
                var capSize = _rand.Next(3, 6);
                return new Action(() =>
                {
                    Water -= 0.4f;
                    Energy -= 0.4f;
                    Grid.Set(vector2.X, vector2.Y - 1, new Cap() { 
                        StalkVector2 = vector2, 
                        MaxDistance = new Vector2I(1,0).DistanceSquaredTo(new Vector2I(capSize,0)),
                        Water = 0.4f,
                        Energy = 0.4f
                    });
                });
            }
        }

        if (up is Stalk stalk && Water > 0.2f && Energy > 0.2f)
        {
            return new Action(() =>
            {
                stalk.Water += 0.1f;
                stalk.Energy += 0.1f;
                
                Water -= 0.1f;
                Energy -= 0.1f;
            });
        }

        return null;
    }

    public Color GetColor(Vector2I vector2) => new Color(0.97f, 0.97f, 0.97f);
    public char Symbol { get; } = '#';
}